core/src/main/java/org/pentaho/platform/workitem/WorkItemLifecycleEvent.java

```diff
@@ -0,0 +1,270 @@
+      logger.error( uhe.getLocalizedMessage() );
+    }
+  }
+
+  /**
+   * Default constructor, needed for serialization purposes.
+   */
+  protected WorkItemLifecycleEvent() {
+  }
+
+  /**
+   * Creates the {@link WorkItemLifecycleEvent} with all the required parameters. The {@code sourceTimestmap} is
+   * set to the current date.
+   *
+   * @param workItemUid     a {@link String} containing unique identifier for the {@link WorkItemLifecycleEvent}
+   * @param workItemDetails a {@link String} containing details of the {@link WorkItemLifecycleEvent}
+   */
+  public WorkItemLifecycleEvent( final String workItemUid, final String workItemDetails ) {
+    this( workItemUid, workItemDetails, new Date() );
+  }
+
+  /**
+   * Creates the {@link WorkItemLifecycleEvent} with all the required parameters.
+   *
+   * @param workItemUid     a {@link String} containing unique identifier for the {@link WorkItemLifecycleEvent}
+   * @param workItemDetails a {@link String} containing details of the {@link WorkItemLifecycleEvent}
+   * @param sourceTimestamp a {@link Date} representing the time the lifecycle change occured.
+   */
+
+  public WorkItemLifecycleEvent( final String workItemUid, final String workItemDetails, final Date sourceTimestamp ) {
+    this.workItemUid = workItemUid;
+    this.workItemDetails = workItemDetails;
+    this.sourceTimestamp = sourceTimestamp;
+
+    // if the workItemUid is null, generate it
+    if ( this.workItemUid == null ) {
+      this.workItemUid = generateWorkItemId();
+    }
+    // Set sourceTimestamp to current date only if not already provided
+    if ( this.sourceTimestamp == null ) {
+      this.sourceTimestamp = new Date();
+    }
+
+    // set the default values for host name and ip, they can be changed directly if needed
+    this.sourceHostName = HOST_NAME;
+    this.sourceHostIp = HOST_IP;
+  }
+
+  public static String generateWorkItemId() {
+    return String.format( ActionUtil.REQUEST_ID_FORMAT, UUID.randomUUID().toString() );
+  }
+
+  /**
+   * Adds the {@code workItemUid} to the given {@link Map}.
+   *
+   * @param map the {@link Map} to which the {@code workItemUid} is being added
+   */
+  public void addUidToMap( final Map map ) {
+    if ( map != null ) {
+      map.put( ActionUtil.REQUEST_ID, getWorkItemUid() );
+    }
+  }
+
+  /**
+   * Looks up the {@code ActionUtil.REQUEST_ID} within the {@link Map}. If available, the value is returned,
+   * otherwise a new uid is generated.
+   *
+   * @param map a {@link Map} that may contain the {@code ActionUtil.REQUEST_ID}
+   * @return {@code ActionUtil.REQUEST_ID} from the {@link Map} or a new uid
+   */
+  public static String getUidFromMap( final Map map ) {
+    String workItemUid = null;
+    if ( map == null ) {
+      workItemUid = generateWorkItemId();
+    } else {
+      workItemUid = (String) map.get( ActionUtil.REQUEST_ID );
+      if ( workItemUid == null ) {
+        workItemUid = generateWorkItemId();
```

Comment:
I was referring to the case when the Map did not have the value and it gets generated.  You have to make sure the caller of this method puts the value in the Map or else the next time it is requested a new value will be returned.